File: python/save_data.py
"""
Load a save file, read/edit named fields, write it back out - with the
platform header and (where supported) checksum handled automatically.

    >>> sd = SaveData.load("reference/pc/ffx2_001", Game.FFX2, "pc")
    >>> sd.get("gil")
    1234
    >>> sd.set("gil", 999999)
    >>> sd.save("output/edited/ffx2_001")
"""

from checksum import Game, update_checksum
from fields import read_field, write_field
from fields_ffx import FFX_FIELDS_BY_NAME
from fields_ffx2 import FFX2_FIELDS_BY_NAME

# Confirmed against real hardware for both games (see README/CHANGELOG):
# Switch prepends an 8-byte header; PC and Vita saves have none.
HEADER_LEN = {"pc": 0, "vita": 0, "switch": 8}

_FIELDS_BY_GAME = {Game.FFX: FFX_FIELDS_BY_NAME, Game.FFX2: FFX2_FIELDS_BY_NAME}

# Both games' checksum algorithms are validated (see checksum.py).
_CHECKSUM_SUPPORTED = {Game.FFX: True, Game.FFX2: True}


class UnknownFieldError(KeyError):
    pass


def coerce_value(raw, current):
    """Parses a user-supplied string `raw` into the same type as `current`
    (as returned by SaveData.get) - bool/float/int."""
    if isinstance(raw, (bool, int, float)):
        raw = str(raw)
    if isinstance(current, bool):
        return raw.strip().lower() in ("1", "true", "yes", "on")
    if isinstance(current, float):
        return float(raw)
    return int(raw)


class SaveData:
    def __init__(self, raw: bytes, game: Game, platform: str):
        if platform not in HEADER_LEN:
            raise ValueError(f"platform must be one of {sorted(HEADER_LEN)}, got {platform!r}")
        self.game = game
        self.platform = platform
        self.header_len = HEADER_LEN[platform]
        self.data = bytearray(raw)
        self._fields = _FIELDS_BY_GAME[game]

    @classmethod
    def load(cls, path, game: Game, platform: str):
        with open(path, "rb") as f:
            return cls(f.read(), game, platform)

    def _field(self, name):
        try:
            return self._fields[name]
        except KeyError:
            raise UnknownFieldError(
                f"no field {name!r} for {self.game.name} "
                f"(known fields: {len(self._fields)})"
            ) from None

    def get(self, name):
        return read_field(self.data, self._field(name), self.header_len)

    def set(self, name, value):
        write_field(self.data, self._field(name), value, self.header_len)

    def fields(self):
        """Returns {name: current_value} for every known field."""
        return {name: self.get(name) for name in self._fields}

    def export_bytes(self, recompute_checksum=True):
        data = bytes(self.data)
        if recompute_checksum and _CHECKSUM_SUPPORTED[self.game]:
            data = update_checksum(data, self.game, self.header_len)
        return data

    def save(self, path, recompute_checksum=True):
        with open(path, "wb") as f:
            f.write(self.export_bytes(recompute_checksum))

    def checksum_supported(self):
        return _CHECKSUM_SUPPORTED[self.game]

File: typescript/src/lib/saveData.ts
// Load a save's bytes, read/edit named fields, export it back out - with
// the platform header and checksum handled automatically. A straight port
// of the Python project's save_data.py.

import { Game, updateChecksum } from "./checksum";
import { readField, writeField, type FieldDef } from "./fields";
import { FFX_FIELDS } from "./generated/fieldsFfx";
import { FFX2_FIELDS } from "./generated/fieldsFfx2";

export type Platform = "pc" | "vita" | "switch";

// Confirmed against real hardware for both games: Switch prepends an
// 8-byte header; PC and Vita saves have none.
export const HEADER_LEN: Record<Platform, number> = { pc: 0, vita: 0, switch: 8 };

// Both games' checksum algorithms are validated (see checksum.ts).
const CHECKSUM_SUPPORTED: Record<Game, boolean> = { [Game.FFX]: true, [Game.FFX2]: true };

function buildIndex(fields: FieldDef[]): Map<string, FieldDef> {
  const m = new Map<string, FieldDef>();
  for (const f of fields) m.set(f.name, f);
  return m;
}

const FIELDS_BY_GAME: Record<Game, FieldDef[]> = {
  [Game.FFX]: FFX_FIELDS,
  [Game.FFX2]: FFX2_FIELDS,
};

const FIELDS_INDEX: Record<Game, Map<string, FieldDef>> = {
  [Game.FFX]: buildIndex(FFX_FIELDS),
  [Game.FFX2]: buildIndex(FFX2_FIELDS),
};

export class UnknownFieldError extends Error {}

export class SaveData {
  readonly game: Game;
  readonly platform: Platform;
  data: Uint8Array;
  readonly headerLen: number;
  private readonly fieldsIndex: Map<string, FieldDef>;

  constructor(game: Game, platform: Platform, raw: Uint8Array) {
    this.game = game;
    this.platform = platform;
    this.data = new Uint8Array(raw); // own copy - never aliases the caller's buffer
    this.headerLen = HEADER_LEN[platform];
    this.fieldsIndex = FIELDS_INDEX[game];
  }

  fieldList(): FieldDef[] {
    return FIELDS_BY_GAME[this.game];
  }

  private field(name: string): FieldDef {
    const f = this.fieldsIndex.get(name);
    if (!f) {
      throw new UnknownFieldError(`no field ${JSON.stringify(name)} for ${this.game} (known fields: ${this.fieldsIndex.size})`);
    }
    return f;
  }

  get(name: string): number | boolean {
    return readField(this.data, this.field(name), this.headerLen);
  }

  set(name: string, value: number | boolean): void {
    writeField(this.data, this.field(name), value, this.headerLen);
  }

  /** Returns {name: current_value} for every known field. */
  fields(): Record<string, number | boolean> {
    const out: Record<string, number | boolean> = {};
    for (const name of this.fieldsIndex.keys()) out[name] = this.get(name);
    return out;
  }

  exportBytes(recomputeChecksum = true): Uint8Array {
    if (recomputeChecksum && CHECKSUM_SUPPORTED[this.game]) {
      return updateChecksum(this.data, this.game, this.headerLen);
    }
    return new Uint8Array(this.data);
  }

  checksumSupported(): boolean {
    return CHECKSUM_SUPPORTED[this.game];
  }
}
